plugin/src/main/java/org/wildfly/plugins/bootablejar/maven/goals/DevWatchContext.java

```diff
@@ -0,0 +1,522 @@
+            }
+            return null;
+        }
+
+        @Override
+        protected Path getInDeploymentPath(Path absolutePath) {
+            if (absolutePath.startsWith(webAppDir)) {
+                return toDeploymentPath(absolutePath, webAppDir.relativize(absolutePath));
+            } else {
+                return super.getInDeploymentPath(absolutePath);
+            }
+        }
+
+        @Override
+        Path getResourcesPath() {
+            return Paths.get("WEB-INF").resolve("classes");
+        }
+
+    }
+    private final Map<WatchKey, Path> watchedDirectories = new HashMap<>();
+    private final Path javaDir;
+    private final Path webAppDir;
+    private final Set<Path> resourceDirectories = new HashSet<>();
+    private final Set<Path> extraDirectories = new HashSet<>();
+    private final Set<Path> cliFiles = new HashSet<>();
+    private final Path pom;
+    private final WatchService watcher;
+    private final Path projectBuildDir;
+    private final Path artifactFile;
+    private final boolean isWebApp;
+    private final boolean isJarApp;
+    private final String fileName;
+    private final Path targetDir;
+    private final ProjectContext ctx;
+
+    DevWatchContext(ProjectContext ctx,
+            WatchService watcher) throws IOException, MojoExecutionException {
+        this.watcher = watcher;
+        this.ctx = ctx;
+        this.projectBuildDir = ctx.getProjectBuildDir();
+        Path mainDir = ctx.getSourceDir().resolve("main");
+        this.javaDir = ctx.getJavaDir();
+        webAppDir = mainDir.resolve("webapp");
+
+        String finalName = ctx.getFinalName();
+        artifactFile = projectBuildDir.resolve(finalName + "." + ctx.getPackaging());
+        if (!Files.exists(ctx.getDeploymentsDir())) {
+            Files.createDirectories(ctx.getDeploymentsDir());
+        }
+        String fName = artifactFile.getFileName().toString();
+        boolean webapp = false;
+        boolean jarapp = false;
+        if (ctx.getPackaging().equals(WAR) || fName.endsWith(WAR)) {
+            if (ctx.isContextRoot()) {
+                fName = "ROOT." + WAR;
+            }
+            webapp = true;
+        } else {
+            if (ctx.getPackaging().equals(JAR) || fName.endsWith(JAR)) {
+                jarapp = true;
+            } else {
+                throw new RuntimeException("Not supported packaging : " + ctx.getPackaging());
+            }
+        }
+        isWebApp = webapp;
+        isJarApp = jarapp;
+        fileName = fName;
+        targetDir = ctx.getDeploymentsDir().resolve(fileName);
+
+        registerDir(ctx.getBaseDir());
+
+        pom = ctx.getPomFile();
+
+        for (Path p : ctx.getResources()) {
+            if (!p.isAbsolute()) {
+                p = ctx.getBaseDir().resolve(p);
+            }
+            // We must add it even if it doesn't exist.
+            // That the way to know the resource files in case they are created later.
```

Comment:
Very minor comment, but this might need to be reworded a bit. Maybe "That way we know the resources files in case they are created later.".

Reply:
Ok, thank-you, yes.